oxidation/libparsec/crates/serialization_format/src/protocol.rs

```diff
@@ -0,0 +1,739 @@
+                                ::rmp_serde::from_slice::<Self>(buf)
+                                    .or_else(|err| {
+                                        // Due to how Serde handles variant discriminant, we cannot express unknown
+                                        // status as a default case in the main schema.
+                                        // Instead we have this additional deserialization attempt fallback.
+                                        let data = ::rmp_serde::from_slice::<UnknownStatus>(buf)?;
+
+                                        match data.status.as_str() {
+                                            #(#known_rep_statuses => Err(err),)*
+                                            _ => Ok(Self::UnknownStatus {
+                                                unknown_status: data.status,
+                                                reason: data.reason,
+                                            })
+                                        }
+                                    })
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+    }
+}
+
+fn quote_custom_enum(
+    name: &str,
+    variants: &[GenCmdNestedTypeVariant],
+    discriminant_field: &Option<String>,
+) -> TokenStream {
+    let name = format_ident!("{}", name);
+    let variants: Vec<TokenStream> = variants
+        .iter()
+        .map(|variant| {
+            let variant_name = format_ident!("{}", variant.name);
+            let discriminant_value = &variant.discriminant_value;
+            match &variant.fields {
+                None => {
+                    quote! {
+                        #[serde(rename = #discriminant_value)]
+                        #variant_name
+                    }
+                }
+                Some(variant_fields) => {
+                    let variant_fields = quote_cmd_fields(variant_fields, false);
+                    quote! {
+                        #[serde(rename = #discriminant_value)]
+                        #variant_name {
+                            #(#variant_fields),*
+                        }
+                    }
+                }
+            }
+        })
+        .collect();
+    match discriminant_field {
+        Some(discriminant_field) => {
+            quote! {
+                #[::serde_with::serde_as]
+                #[derive(Debug, Clone, ::serde::Deserialize, ::serde::Serialize, PartialEq, Eq)]
+                #[serde(tag = #discriminant_field)]
+                pub enum #name {
+                    #(#variants),*
+                }
+            }
+        }
+        None => {
+            quote! {
+                #[::serde_with::serde_as]
+                #[derive(Debug, Clone, ::serde::Deserialize, ::serde::Serialize, PartialEq, Eq)]
+                pub enum #name {
+                    #(#variants),*
+                }
+            }
+        }
+    }
```

Comment:
`quote` is able to expand rust option to nothing, so you could factorize that code

Reply:
`discriminant_field` is not an iterator so I cannot do something like `#(#[serde(tag = #discriminant_field)])*`... do you have something else in mind ?